### DISCUSSION TITLE
pyplot figures not opening or saving to disk

### DISCUSSION THREAD

**Discussion by ghylander** [pyinstaller/pyinstaller#7286] [Help] [answered]
This seems to be a recurrent issue.
I tried to place the `libiomp5md.dll` library in the same path as the output exe.
I'm using python 3.11

**Short explanation and answer:**
Together with matplotlib, I was using mpld3 too. This library declares `matplotlib.use('Agg')` internally, and was causing PyInstaller to only select that backend. Other parts of the code which did not use mpld3 used a different backend (Qt5Agg), but because it was not selected by PyInstaller, the 'compiled' application did not show or save those figures.

If this is your issue, in your code simply declare `matplotlib.use('Qt5Agg')`, PyInstaller will pick up both Agg and Qt5Agg backends.

**Comment by ghylander** [comment]
I also added the lines
```
from multiprocessing import freeze_support
freezee_support()
```
In case there was any issue with matplotlib not blocking the thread

What happens is not that the program fails to plot or save the figures and exits, it just seems to ignore the `pyplot.show()` or `pyplot.savefig()` lines, everything else seems to execute with no problem

**Comment by rokm** [comment, marked as answer]
Can you provide an example that doesn't work when frozen? The basic
```python
import numpy as np
import matplotlib.pyplot as plt

x = np.arange(0, 5, 0.1)
y = np.sin(x)
plt.plot(x, y)
plt.show()
```
seems to work as expected (python 3.11.0 from python.org, matplotlib 3.6.2 and pyinstaller 5.6.2 installed via pip).

Are you using anaconda python (since you mention `libiomp5md.dll`...)? 

Are you using or trying to use specific matplotlib backend? In that case, do you see PyInstaller collecting that backend during the build or not (the message that looks like: `INFO: Selected matplotlib backends: ['TkAgg']`).

**Reply by ghylander** [reply]
I am using a conda environment which includes only the bare needed to run the application
I am not specifying any matplotlib backend, PyInstaller reports the following: 
```
64631 INFO: Matplotlib backend selection method: automatic discovery of used backends
64758 INFO: Selected matplotlib backends: ['Agg']
```

The full source of the code is quite large, I'll try to simplify it and provide a standalone code.
Currently, my application looks like this:

```
root
. . . main.py
. . . module
. . . . . . __init__.py
. . . . . . submodule1.py
. . . . . . submodule2.py
. . . . . . submodule3.py
. . . . . . submodule4.py
. . . . . . submodule5.py
```

All `plot()` and `show()` are carried out in any single submodule
I'm also using matplotlib 3d projections, but 2d projections don't seem to work either

**Reply by rokm** [reply]
> ```
> 64631 INFO: Matplotlib backend selection method: automatic discovery of used backends
> 64758 INFO: Selected matplotlib backends: ['Agg']
> ```

I suppose this means that somewhere in your code, you have a `matplotlib.use('Agg')`, which causes PyInstaller to collect only that backend without trying to import and collect the default UI backends (such as `TkAgg`).

Take a look at: https://pyinstaller.org/en/stable/hooks-config.html#matplotlib-hooks

And try adding the `hooksconfig` dict to `Analysis` in the .spec file. In that dict, add `matplotlib` dict with `backends`, which you can try setting either to "all" (since you have a minimal environment) or to list of specific backends that you want collected (e.g., `['TkAgg', 'Agg']`). Then build the program using the spec file instead of the main.py file.

**Reply by ghylander** [reply]
The issue indeed seems to be the backend attached by PyInstaller
I 'compiled' a single submodule, and PyInstaller reported the following:
```
63013 INFO: Trying determine the default backend as first importable candidate from the list: ['Qt5Agg', 'Gtk3Agg', 'TkAgg', 'WxAgg']
63799 INFO: Selected matplotlib backends: ['Qt5Agg']
```
After adding `matplotlib.use('Qt5Agg')` to that module, when calling to the module from the 'compiled' application figures are shown and saved normally.

I do not have anywhere in the code explicitly declared `matplotlib.use('Agg')` (or any other backend)
I am however using matplotlib animations, using ffmpeg as backend to export the animations:

`plt.rcParams['animation.ffmpeg_path']='path/to/ffmpeg/binary`

The ffmpeg binary itself is bundled inside the application.

I also subclass matplotlib's ffmpegwriter with a more optimized one:

```
from matplotlib.animation import FFMpegWriter

class FasterFFMpegWriter(FFMpegWriter):
    '''FFMpeg-pipe writer bypassing figure.savefig.'''
    def __init__(self, **kwargs):
        '''Initialize the Writer object and sets the default frame_format.'''
        super().__init__(**kwargs)
        self.frame_format = 'argb'

    def grab_frame(self, **savefig_kwargs):
        '''Grab the image information from the figure and save as a movie frame.

        Doesn't use savefig to be faster: savefig_kwargs will be ignored.
        '''
        try:
            # re-adjust the figure size and dpi in case it has been changed by the
            # user.  We must ensure that every frame is the same size or
            # the movie will not save correctly.
            self.fig.set_size_inches(self._w, self._h)
            self.fig.set_dpi(self.dpi)
            # Draw and save the frame as an argb string to the pipe sink
            self.fig.canvas.draw()
            self._proc.stdin.write(self.fig.canvas.tostring_argb())
        except (RuntimeError, IOError) as e:
            out, err = self._proc.communicate()
            raise IOError('Error saving animation to file (cause: {0}) '
                      'Stdout: {1} StdError: {2}. It may help to re-run '
                      'with --verbose-debug.'.format(e, out, err)) 

anim = animation.FuncAnimation(fig,
                               GraphicsAnimateFunc,
                               frames = len(Time)//20,
                               init_func = init,
                               interval = 1,
                               blit = True,
                              )
            
writer = FasterFFMpegWriter(fps=30)
anim.save("Graphics.mp4",
          writer=writer,
          dpi=200,
         )
```




Perhaps this is interfering with the backend estimation, or doing some stuff under the hood with `matplotlib.use()`?

**Reply by rokm** [reply]
Yeah, it could be a 3rd party module calling `matplotlib.use()`. If you want to check, try modifying the `hook-matplotlib.backends.py` in your local PyInstaller installation; replace this line:

https://github.com/pyinstaller/pyinstaller/blob/e1ace295f7b55c31c0b831792a9097cdf1cc6672/PyInstaller/hooks/hook-matplotlib.backends.py#L101

with

```python
        tmp_backends = _recursive_scan_code_objects_for_mpl_use(co)
        if tmp_backends:
            logger.info("Found backends via matplotlib.use() in %r: %r", name, tmp_backends)
            used_backends += tmp_backends
```

**Reply by ghylander** [reply]
Spot on, the mpld3 library is indeed declaring `matplotlib.use('Agg')`:
```
75844 INFO: Matplotlib backend selection method: automatic discovery of used backends
75894 INFO: Found backends via matplotlib.use() in 'mpld3': ['Agg']
75987 INFO: Found backends via matplotlib.use() in 'submoduleX': ['Qt5Agg']
75987 INFO: Selected matplotlib backends: ['Agg', 'Qt5Agg']
```
Thanks! I think I would have taken quite long to realise it was a matplotlib backend issue had it not been for you asking if I was specifying a backend
